src/http-gateways/gateway-detection.md

```diff
@@ -0,0 +1,102 @@
+---
+title: Detecting User-Preferred IPFS Gateway
+description: >
+  Specification of the rules and standards for detecting and identifying
+  user-preferred IPFS gateways within applications, enabling seamless
+  integration and user control.
+date: 2023-10-03
+maturity: reliable
+editors:
+  - name: Mark Gaiser
+    github: markg85
+  - name: Marcin Rataj
+    github: lidel
+    url: https://lidel.org/
+    affiliation:
+        name: Protocol Labs
+        url: https://protocol.ai/
+tags: ['httpGateways', 'integratingHttpGateways']
+order: 99
+---
+
+
+## Introduction
+
+This document defines conventions for how applications can identify available
+IPFS gateway, and how IPFS gateway implementations can signal own endpoint to
+client applications.
+
+## Specification
+
+There are two ways of hinting user-prefered gateway URL:
+
+- Setting the `IPFS_GATEWAY` environment variable
+- Creating a `gateway` file at a well-known path
+
+Applications SHOULD evaluate these hints in order and stop on the first match:
+
+1. Check if a valid `IPFS_GATEWAY` environment variable is set
+2. Check if a valid `gateway` file is present at one of well-known filesystem paths
+
+
+### `IPFS_GATEWAY` Environment Variable
+
+When `IPFS_GATEWAY` environment variable is set, the value MUST be interpreted
+as URL of IPFS Gateway application to use.
+
+This variable SHOULD override gateway selection done by all other means, including
+internal application configuration.
+
+### The `gateway` Configuration File
+
+Client application SHOULD check if file is present at specific filesystem paths, in order:
+
+1. If `IPFS_PATH` is set, try `$IPFS_PATH/gateway`
+2. If `HOME` is set, try `$HOME/.ipfs/gateway`
+3. If `$XDG_CONFIG_HOME` is set, try `$XDG_CONFIG_HOME/ipfs/gateway`
+4. If `/etc` exists, try `/etc/ipfs/gateway`
+5. Try OS-specific paths
+   - Windows
+     1. `%LOCALAPPDATA%/ipfs/gateway` (local user)
+     2. `%APPDATA%/ipfs/gateway` (roaming user)
+     3. `%PROGRAMDATA%/ipfs/gateway` (global)
+   - macOS
+     1. `$HOME/Library/Application Support/ipfs/gateway` (user)
+     2. `/Library/Application Support/ipfs/gateway` (global)
+   - Linux
+     1. `$HOME/.config/ipfs/gateway` (user)
+     2. `/etc/ipfs/gateway` (global)
+
+When `gateway` file is present, the file contents MUST be interpreted as line
+separated (`\n` or `\r\n`) `text/plain` file.
+
+The first line of `gateway` file MUST be a valid `http://` or `https://` URL.
+
+Implementations that only need one URL SHOULD use the first one and ignore the rest of the file.
```

Comment:
@markg85 I wrote this bit naively, and  now I realized that having more than one URL will cause issues in curl ([code](https://github.com/markg85/curl/blob/22e2eaa278b7be853506bc3e81d349736a147140/src/tool_operate.c#L745-L757)). 

Any concerns with simplifying the spec and stating that only one URL is allowed in the file?

Reply:
Ohh ouch!

For this spec. I'd keep it at a first line requirement but not limited to 1 line. Aka, it's fine as-is. 

Curl needs some attention to catch the multiline scenario regardless of this spec. As just any other data after the first line is gonna break it.

Reply:
Ok, my understanding is this problem was solved in https://github.com/curl/curl/commit/859e88f6533f9e1f890f6f39f34178871c48869e#diff-1fa338b6a8d4b9a9ffd2243527ac4951c2576cb32dc28440f36f4f59e60643e2R785, the `gateway` file can now have more than one line, curl ignores everything after the first one.